### PR TITLE
feat: activity log

### PR DESCRIPTION
**Pull request by tamassoltesz**
## Summary of change

Aggregate landing of the **full** activity-log feature into the `12.2` version
branch. This started as the three foundational units and has since grown to the
complete feature; the whole thing is collected here so it releases as one
coherent unit. It contains:

**Foundational units**
- **Lifecycle-event vocabulary + JSON payload schema**
  (`io.supertokens.auditlog.lifecycle`) — a closed, versioned schema with a
  builder/validator (`schemaVersion`, parse/validate split). (#1381)
- **Compile-time audit-transaction guard** — `AuditEnforcementAspect` plus the
  `@UnauditedTransaction` allowlist and a shrink-only baseline test
  (`UNAUDITED_TRANSACTION_BASELINE = 57`, matching the current annotation count
  after the net conversions below). (#1382)
- **In-memory (SQLite) storage parity** for the activity-log/rollup contract —
  the transactional audit insert (`createActivityLogEntry_Transaction`), the
  fold/reconcile rollup (`rollupLastActiveFromActivityLog_Transaction`), the
  cheap `hasUnfoldedActivitySince` signal, and `maintainActivityLogPartitions`
  (a documented no-op in-memory — see the interface javadoc). (#1395)

**Feature units built on the foundation**
- **Lifecycle emit sites** — link/unlink (#1368) and deleteUser /
  user-group-deletion / tenant-association (#1374) now write lifecycle events
  through the audited transaction path.
- **`CountDeltaInterpreter` read-side fold** (#1375) — a pure,
  property-tested fold of the activity log into count deltas, with a burst cap.
- **Last-active rollup cron + dirty signal** (#1369) — `RollupUserLastActive`
  consuming `hasUnfoldedActivitySince` / `rollupLastActiveFromActivityLog_Transaction`;
  both crons are wired in `Main.init`.
- **`user_last_active` cutover** (#1371) — the rollup cron becomes the sole
  writer of `user_last_active`; `updateLastActive()` only appends a
  `user_last_active` activity-log event and marks the storage dirty. The
  count/MAU projection is therefore updated asynchronously (within a rollup
  interval) rather than synchronously — an accepted tradeoff documented at the
  emit sites.

Part of PLAN-010 / PLAN-011 / PLAN-012.

## Related issues

Aggregates the following already-reviewed units (each merged/approved as its own
PR before landing here): #1381, #1382, #1395, #1368, #1374, #1375, #1369, #1371.

## Cross-repo dependency

The activity-log storage contract lives on the plugin-interface `feat/activity-log`
branch (resolved by branch name; `pluginInterfaceSupported.json` → `10.0`). The
PostgreSQL plugin's matching implementation is on its own `feat/activity-log`
branch (umbrella PR postgresql-plugin#387). One companion fix is required for the
postgres build to compile against the new PI contract:
supertokens/supertokens-postgresql-plugin#394 (aligns the three pre-existing
connection-taking txn variants to `(TenantIdentifier, TransactionConnection, …)`
— the postgres mirror of in-memory commit `0b8f0105`). Until #394 lands on
postgres `feat/activity-log`, `Tests (postgresql)` here stays red at
`:supertokens-postgresql-plugin:compileAspectj`.

## Test Plan

Each unit shipped with its own tests, which run as part of this branch's CI
(`Tests (sqlite)` green; `Tests (postgresql)` blocked only by the compile
dependency above):

- **Payload schema** (#1381): round-trip serialization + an exhaustive rejection
  matrix for the closed schema (`schemaVersion`, unknown fields, malformed
  payloads).
- **Audit guard** (#1382): the shrink-only baseline test — fails if any
  transaction opens outside the audited path or the allowlist grows past the
  baseline (now 57; the source-scan runs under `Tests (sqlite)`, no DB).
- **In-memory rollup** (#1395): `ActivityLogRollupTest` covers fold idempotency,
  monotonicity (`MAX` merge), reconcile, and transactional atomicity of the
  audit insert; `ActivityLogRetentionTest` asserts the in-memory
  `maintainActivityLogPartitions` no-op (partition/retention proper is covered by
  the PostgreSQL plugin's `ActivityLogPartitionTest`).
- **Emit sites** (#1368, #1374): lifecycle events are asserted at link/unlink,
  deleteUser, user-group-deletion, and tenant-association.
- **Interpreter** (#1375): property test over the fold (17/17), including the
  burst cap.
- **Rollup cron + cutover** (#1369, #1371): rollup idempotency/monotonicity and
  the async-projection behaviour (counts/MAU converge within a rollup interval).
  Note: `TestAppData` excludes `user_last_active` from its synchronous
  populate-then-assert table sweep, because after the cutover the projection is
  written asynchronously; app-deletion cleanup of that table remains guaranteed
  by its `ON DELETE CASCADE` FK to `apps` (see the inline reply thread).

## Documentation changes

None required at the core level; the payload schema is documented inline via the
validator and the interface javadoc.

## Checklist for important updates

- [x] Changelog has been updated (per-unit changie/CHANGELOG entries, collected
  under `## [12.2.0]`)
- [ ] `coreDriverInterfaceSupported.json` — no CDI change in this branch
- [x] `pluginInterfaceSupported.json` — bumped to `10.0` for the activity-log
  contract
- [x] Issued against the latest non-released version branch (`12.2`)

### CODE REVIEW

**Comment by tamassoltesz** [comment]
what are you talking about unmerged PRs? https://github.com/supertokens/supertokens-core/pull/1399 here is one PR you state that it's unmerged. It was merged way before your statement. Don't make assumptions, check!

**Comment by tamassoltesz** [comment]
postgresql tests are failing. Inspect those.

**Comment by tamassoltesz** [comment]
> **Separate production concern (flagged, not fixed).** The same FK violation reveals a latent _production_ bug independent of the test reset: because `activity_log` rows are intentionally retained after an app is deleted while `user_last_active` cascades on app delete, the rollup fold will FK-violate in production whenever an app is deleted within the fold window (deleted app's `user_last_active` rows cascade away, its `activity_log` rows remain, next fold re-inserts them). No current test exercises it (nothing deletes an app then folds), so it doesn't block CI. Suggested follow-up: guard the fold so it never resurrects a row for an app_id absent from `apps` (an `EXISTS`/join against `apps` in the FOLD query, in both this repo's in-memory `ActiveUsersQueries` and the plugin) — I left the approach to you since it changes the reviewed rollup SQL in two repos. Happy to implement it if you'd like.

Yes please, implement the guard. Also, add tests that exercises this path.

File: src/main/java/io/supertokens/inmemorydb/queries/ActiveUsersQueries.java [hunk 1]
```diff
@@ -148,6 +148,49 @@ public static void deleteUserActive_Transaction(Connection con, Start start, App
...
+
+        // SQLite's two-argument max() is the scalar GREATEST, so the upsert stays monotonic.
+        // The apps guard skips activity for apps deleted within the window: activity_log rows are
+        // intentionally retained after an app is deleted (no app_id cascade), but user_last_active
+        // cascades on app delete, so folding a since-deleted app's rows would violate the
+        // user_last_active -> apps foreign key. EXISTS keeps the fold set to still-existing apps only.
+        String FOLD_QUERY = "INSERT INTO " + userLastActiveTable + " (app_id, user_id, last_active_time)"
+                + " SELECT app_id, primary_or_recipe_user_id, MAX(created_at) FROM " + activityLogTable + " al"
+                + " WHERE event_type = 'user_last_active' AND created_at >= ?"
+                + " AND EXISTS (SELECT 1 FROM " + appsTable + " a WHERE a.app_id = al.app_id)"
```
[the remark this answers is not in the document]
Comment (tamassoltesz) [reply src/main/java/io/supertokens/inmemorydb/queries/ActiveUsersQueries.java]: postgresql PR was merged. Please verify the tests are green

**Comment by tamassoltesz** [comment]
Review of the combined feat/activity-log feature set (this PR + supertokens-postgresql-plugin#387 + supertokens-plugin-interface#218), focused on the ledger/fold correctness and cross-repo contracts. Findings that belong to the core side:

---

**1. `CountDeltaInterpreter.TENANT_DISASSOCIATION` over-decrements when a multi-member group stays in the tenant via another member**

`src/main/java/io/supertokens/auditlog/lifecycle/CountDeltaInterpreter.java:201` applies `-1` whenever `groupBefore.tenantIds` contains the tenant. But `Multitenancy.removeUserIdFromTenant` removes only *one member's* tenant mapping; when primary P and linked member R are both mapped to tenant `t` and R is disassociated, the group is still present in `t` via P, so `getUsersCount(t)` (distinct `primary_or_recipe_user_id`) is unchanged — yet the fold applies `-1`. The served anchor+fold `/users/count` (now the default path from CDI 5.6) undercounts by 1 until the next re-anchor, and the shadow audit reports a DISCREPANCY for every such event. `TENANT_ASSOCIATION` has the symmetric guard ("unless already present"); disassociation can't be derived from group-level before-presence alone — the event needs after-presence (like `USER_DELETION`) or a member-level presence list.

**2. The unlink-a-primary-with-members branch emits an `account_unlinking` event that always folds to delta 0, losing the `-1` for tenants the group leaves**

`AuthRecipe.unlinkAccounts`, branch where `lockedRecipeUser.getPrimaryUserId().equals(recipeUserId)` and the group has other members (`AuthRecipe.java` ~line 185-200): `deleteUserHelper` deletes the primary's own login method, and the event emitted is `forAccountUnlinking(remainingGroupAfter, freedMemberAfter = empty)`. The interpreter's `ACCOUNT_UNLINKING` arm is `+1` per tenant in `intersection(remainingAfter, freedAfter)` = `intersection(…, ∅)` = 0 everywhere. But if the deleted login method was the group's only presence in some tenant `t1`, the group actually leaves `t1` and the exact count drops by 1 — the ledger never records it. Permanent overcount in the served count plus recurring shadow-audit discrepancies. This branch is semantically a member deletion; it should emit `user_deletion` with group before/after lists.

**3. Sign-up recency is lost for tenants that live on a different user pool than the app's public tenant**

The `user_creation` lifecycle event is written on the *request tenant's* storage (`SignUpAPI` uses `getTenantStorage(req)`, and `startAuditedTransaction` writes on that connection), while the counted projection is the *app public tenant's* `user_last_active`: `ActiveUsers.countUsersActiveSince` reads `StorageLayer.getStorage(appIdentifier.getAsPublicTenantIdentifier(), …)` (ActiveUsers.java:206). For an app whose tenant `t1` has its own database, the sign-up's `user_creation` lands in pool Y's activity_log and can only ever fold into pool Y's projection — pool X's counted projection never sees it. Before the cutover, `SignUpAPI` called `updateLastActive`, which wrote to the public-tenant storage, so this is a silent regression: a user who signs up on such a tenant and does nothing else is missing from active-user counts entirely (their later `sign_in`/`token_refresh` pings do go to the public pool, which bounds the damage to sign-up-only users). `SignUpAPI`'s `markLastActiveRollupDirty` also marks pool X dirty while the unfolded event sits in pool Y. Same pattern in thirdparty/passwordless/webauthn sign-up.

**4. First-run/catch-up fold can be permanently skipped in multi-node deployments: the advisory-lock loser still advances the watermark**

`RollupUserLastActive.doTaskPerStorage` writes `WATERMARK_KEY = runStart` unconditionally in the same transaction after `rollupLastActiveFromActivityLog_Transaction` returns — but the postgres implementation returns *silently without folding* when another instance holds the `last_active_rollup` advisory lock (pg `ActiveUsersQueries.java:224`). Worst case is exactly the cutover scenario: first-ever pass, watermark null, window = full 31-day retention catch-up. Node A takes the lock and starts the slow catch-up fold; node B loses the lock, folds nothing, and commits `watermark = runStartB`. If A then fails (crash, statement timeout on the big fold), every subsequent window starts at `watermark − 15min` and the historic activity is never folded — `user_last_active` (and MAU/billing counts) silently lose up to 31 days of credit, with no backstop that reaches back (the backstop fold also uses the watermark window). The lock-skip needs to propagate to the caller so the watermark is not advanced, or the watermark write must be conditional on the fold actually having run. (Aggravator: the advisory-lock key is the constant `last_active_rollup`, scoped per physical Postgres database — two user pools configured as different schemas/table-prefixes in the same database contend on one lock, so pool B's fold can be skipped by pool A's while B's watermark still advances.)

**5. The CDI import-user-with-password-hash path marks imported users as active (and never wakes the rollup)**

`EmailPassword.createUserWithPasswordHash` emits `user_creation` — which is in `RollupEventTypes.FOLD_SET` — for users imported via `ImportUserWithPasswordHashAPI`. Bulk import deliberately emits `user_import` instead ("an imported user is present, not active", per the RollupEventTypes javadoc and the core#1403 decision), so the two import routes now disagree: a migration done through the CDI import API stamps every imported user into `user_last_active` at import time, inflating MAU for the first activity window — the precise outcome `user_import` was carved out to prevent. (Secondary: this path also never calls `markLastActiveRollupDirty`, so even its unintended activity credit waits for the backstop pass.) Both import paths should emit `user_import`.

**6. `linkAccounts` captures group presence before any row lock — the pre-lock race `deleteUser` documents and avoids**

In `AuthRecipe.linkAccounts` the two `groupPresenceInTransaction` reads run at the top of the transaction, *before* `linkAccounts_Transaction` acquires its locks. Under READ COMMITTED, a concurrent `addUserIdToTenant`/link committing on either group between the read and the lock makes the recorded before-lists stale; if the missed tenant is one where both groups were present, the `-1` for that tenant is never folded — a permanent overcount. The deleteUser path's own comment (AuthRecipe.java ~718-727) names this exact race as the reason its snapshot is read under the row lock, and `unlinkAccounts`/`Multitenancy` both lock first; `linkAccounts` should too (lock, then read presence, then link).

**7. In-memory storage lost activity-log retention entirely**

`inmemorydb/Start.maintainActivityLogPartitions(int)` is now a pure no-op, but before this branch the in-memory implementation enforced retention with a direct `DELETE` (`ActivityLogQueries.deleteEntriesOlderThanRetention`, now deleted). The caller-supplied `retentionDays` is ignored and `activity_log` grows without bound on the in-memory store — one row per sign-in/refresh per (app,user) per 5 minutes, or per request with `activity_log_throttle_enabled=false`. The core changelog even describes this as "stays a no-op", which doesn't match the removed behavior. The direct delete should be restored with the `retentionDays` parameter.

---

Cross-repo findings posted on the sibling PRs: fold resurrection of deleted users + reconcile-vs-unlink ordering (postgresql-plugin#387), and the duplicated/stale event vocabulary (plugin-interface#218).

**Comment by tamassoltesz** [comment]
please fix conflicts in changelog and make it shorter (significantly!) No one wants to read a novel, disguised as a changelog. Also, this is going to be part of 12.2, so move the related sections under 12.2

**Comment by tamassoltesz** [comment]
3# for the MAU count: the event should count in the MAU _and_ also should be placed in its own tenant's DB too.
4# Yes, go with the cross-repo fixes.
1# Fix it.
2# Agreed, fix it.
6# Fix it.

**Comment by porcellus** [comment]
Decisions on all open findings in one place. TL;DR: #1/#2/#6 green-lit as you proposed, #4 green-lit as the void→result change, #3 gets a different and bigger fix than either option you offered. Sequencing: land the point fixes first (#6, #2, #1, then the cross-repo #4) as small reviewable changes; the #3 rework follows as its own PR.

## #3 — neither proposed routing; we're moving to a per-storage projection

While reviewing this we found the problem is wider than sign-up credit. The fold's `EXISTS (app_id_to_user_id)` guard is storage-local, and a user exists on exactly one storage per app (that's why `findStorageAndUserIdMappingForUser` iterates storages). So on the app-public storage, *every* event credited to a separate-DB tenant's user — including the `sign_in`/`session_create`/`token_refresh` events `updateLastActive` already routes there — fails the guard and is silently dropped. Post-PR those users contribute zero MAU; pre-PR the direct writes counted them. Both of your options feed the same guard, so neither fixes it.

1. **Events stay on the user-pool storage.** `updateLastActive` routes to the request tenant's storage (drop the `getAsPublicTenantIdentifier()` redirect in storage selection). The transactional lifecycle emits are already there.
2. **The fold stays storage-local (unchanged); the read side sums.** `countUsersActiveSince(app)` — and `countUsersThatHaveMoreThanOneLoginMethodOrTOTPEnabledAndActiveSince` — iterate `getStoragesForApp` and sum. A user exists on exactly one storage per app, so no double count.
3. **`markLastActiveRollupDirty` must mark the tenant storage's pool**, not app-public's, or the sign-up promptness nudge wakes the wrong storage's fold.
4. **Drop the `deleteUserActive_Transaction` call from `deleteNonAuthRecipeUser`.** Deliberate semantics change: a user active in the window and then deleted keeps counting for the remainder of the window. (History: MAU shipped in #564 with no scrub; the scrub arrived as an uncommented one-liner in the #555 multitenancy delete-all-rows sweep — data hygiene, not a MAU decision. The activity log already retains richer post-deletion data than this row, so this is consistent with the feature's own retention stance.) The `account_linking` reconcile stays — it prevents one person counting twice, a different job.
5. **Keep the `app_id_to_user_id` guard, rewrite its justification.** With per-storage routing it's correct (events, users, and projection are colocated), its anti-resurrection job disappears with (4), and it becomes misroute insurance: the one tenant-less emit path (`SessionRemoveAPI`'s app-wide sign-out) still writes to app-public, and without the guard that would create a wrong-storage projection row that the summed read double-counts.
6. **One-time migration** All legacy recency lives on app-public. Add migration notes to delete rows whose `(app_id, user_id)` has no local `app_id_to_user_id` row — storage-local SQL, a no-op for single-storage installs. Transient undercount for cross-storage users until their next activity, which beats the double-count of doing nothing. Do NOT make this a recurring scrub: after (4), a deleted local user's retained row is indistinguishable from a cross-storage user's row.
7. **Failing-first test that pins the whole class:** two-storage app, sign in on the separate-DB tenant, run the rollup, assert `countUsersActiveSince(app) == 1`. Passes on master, fails on this branch today.

## #4 — approved: void → fold-happened result, across the three repos

Advance the watermark only when a fold actually happened, and re-arm the dirty flag on a skipped fold so the next tick retries. Please reject the blocking-lock alternative explicitly in the PR description (it holds a pool connection through catch-up folds — exactly when they're long). The loss case to cover with a test: null-watermark catch-up where the lock loser commits `watermark = now` and the winner rolls back — the deep window is then never re-folded, since later windows only reach back the 15-minute margin.

## #1 — approved as proposed

After-removal presence in the `tenant_disassociation` payload, computed at the `Multitenancy.removeUserIdFromTenant` emit site; the interpreter decrements only when the group was present before and absent after (same pattern as `user_deletion`). Compute the after-presence under the existing lock ordering — and while you're there, check whether locking only the one user gives a consistent `groupBefore`, since that's the same staleness class as #6.

## #2 — approved as proposed

Emit `user_deletion` (with before/after presence) for the deleted member in the unlink-primary-with-members branch, instead of `account_unlinking` with empty `freedMemberAfter` — that interpreter arm can only ever produce `+1`s, so it structurally cannot express the `-1`. The test should pin that exactly one deletion event comes out of that path (`deleteUserHelper` itself doesn't emit).

## #6 — approved as proposed

Lock first, then read presence, matching `deleteUser`/unlink/`Multitenancy`.

## #7 — agreed, no change.

Each point fix with a failing-first test, as you proposed.

(The three issue links at the bottom of your message were auto-linkified against ancient unrelated issues — ignore them.)

File: src/main/java/io/supertokens/emailpassword/EmailPassword.java [hunk 2]
```diff
@@ -118,45 +122,55 @@ public static AuthRecipeUserInfo signUp(TenantIdentifier tenantIdentifier, Stora
...
+                                hashedPassword, timeJoined);
+                        if (isFakeEmail) {
+                            // A fake email is treated as pre-verified; do it on the same connection so the user,
+                            // its verification and the user_creation event all commit (or roll back) together.
+                            evStorage.updateIsEmailVerified_Transaction(appIdentifier, con,
+                                    user.getSupertokensUserId(), email, true);
+                            user.loginMethods[0].setVerified(); // newly created user has only one loginMethod
                         }
-                        throw new StorageQueryException(e);
+                        AuditLogEvent event = LifecycleAuditEvent.forUserCreation(appIdentifier,
```
[the remark this answers is not in the document]
Comment (tamassoltesz) [reply src/main/java/io/supertokens/emailpassword/EmailPassword.java:150]: got your replies from @porcellus. Go ahead and implement the fixes!

File: src/main/java/io/supertokens/cronjobs/rollupUserLastActive/RollupUserLastActive.java [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+            final long windowStartMillis = windowStart;
+            AppIdentifier appIdentifier = representative.toAppIdentifier();
+            // The rollup consumes the activity log to maintain a derived projection; it emits no audit
+            // events of its own, so it runs through the audited-transaction combinator's zero-event path.
+            // The combinator owns the commit.
+            activityLogStorage.startAuditedTransaction(appIdentifier, con -> {
+                activeUsersStorage.rollupLastActiveFromActivityLog_Transaction(con, windowStartMillis);
+                // Advance the watermark to the run-start time (not "now after the fold"): the next window
+                // reaches back to windowStart = watermark - MARGIN, so under-advancing is always safe.
+                activeUsersStorage.setKeyValue_Transaction(representative, con, WATERMARK_KEY,
```
[the remark this answers is not in the document]
Comment (tamassoltesz) [reply src/main/java/io/supertokens/cronjobs/rollupUserLastActive/RollupUserLastActive.java]: the PR was merged, please proceed

File: src/main/java/io/supertokens/emailpassword/EmailPassword.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (tamassoltesz) [reply src/main/java/io/supertokens/emailpassword/EmailPassword.java:150]: please proceed with it.

**Comment by tamassoltesz** [comment]
## Intensive review — correctness / security / readability-SOLID (core findings)

Reviewed the `feat/activity-log` tip (`a22601ed`) together with the plugin-interface (#218) and postgresql-plugin (#387) companions, against the 12.2 base. The feature hangs together well; the security lens came back clean (SQL parameterized or fixed-literal, JSONB rejects malformed payloads at write time, tenant/app isolation holds on the paths reviewed). Findings that land in core:

### 1. CORRECTNESS — CONFIRMED: in-memory rollup reconcile diverges from Postgres
`src/main/java/io/supertokens/inmemorydb/queries/ActiveUsersQueries.java:197`

The in-memory `RECONCILE_QUERY` lacks the ordering guard the Postgres reconcile has (`AND al.created_at >= ula.last_active_time`, pg `ActiveUsersQueries.java:276`). So the two are **not** "result-identical" as the javadoc at line 165 claims.

**Failure:** a recipe user is linked (`account_linking` at `t_link`), then unlinked, then active again (`t_active > t_link`), all within one rollup window. The fold sets the row's `last_active_time = t_active`; Postgres keeps it (`t_link < t_active` fails the guard) but the in-memory reconcile deletes it, because it only checks that an `account_linking` event exists with `created_at >= windowStart`. The still-active user is dropped from `user_last_active` and undercounted in MAU/active-user reads on SQLite storage. `ActivityLogRollupParityTest` passes only because it does not exercise the reactivate-after-unlink case.

**Fix:** add `AND al.created_at >= <userLastActiveTable>.last_active_time` to the in-memory `EXISTS` sub-select, and add a parity-test case for link→unlink→reactivate-in-window. Not addressed by #1410.

### 2. CORRECTNESS / SOLID — CONFIRMED drift hazard: event-type vocabulary is triplicated
`src/main/java/io/supertokens/auditlog/lifecycle/LifecycleEventType.java`

plugin-interface #219 (delivered by #220) added the shared vocabulary classes, and core already consumes the shared `RollupEventTypes` and `ActivityEventType`. But the consolidation is only partial, so the drift hazard persists in two spots:
1. Core still keeps its own `src/main/java/io/supertokens/auditlog/lifecycle/LifecycleEventType.java` — a duplicate of the plugin-interface `LifecycleEventType` that was never retired.
2. The reconcile queries in both storages still hardcode the `'account_linking'` literal rather than reading it from the shared source.

Rename a constant or add a type in one copy and the fold silently credits nobody, or the reconcile stops firing — the projection quietly stops advancing, with no error. Closing this out means retiring core's `LifecycleEventType` duplicate (consume the plugin-interface enum) and pulling the reconcile-trigger literal under the shared source too — it lives inside SQL string concatenation, so it's the copy most likely to be missed.

### 4. EFFICIENCY / ALTITUDE — PLAUSIBLE: burst cap uses an app-wide window but serves per-tenant
`src/main/java/io/supertokens/authRecipe/ApproximateUserCount.java:174`

`getActivityLogEntriesForApp` filters by `app_id`, not tenant, but serving is per-tenant. One tenant's bulk import (or mass delete) producing > `DEFAULT_BURST_CAP` (10,000) lifecycle events in the anchor window trips the burst branch for **every other tenant of the app**, each of which then calls `computeAnchor → getUsersCount` synchronously on the serving thread — the exact O(N) recompute the anchor+fold design exists to avoid. A bulk op on one tenant becomes an app-wide latency cliff on `GET /users/count`.

### 5. CORRECTNESS — PLAUSIBLE (low / currently unreachable): `approximate` flag inverted on the non-ActivityLogStorage fallback
`src/main/java/io/supertokens/authRecipe/ApproximateUserCount.java:149`

The fallback returns `ApproximateCountResult(exactCount, approximate=true, now)` — it serves an **exact** count but sets `approximate=true`, contradicting the field's contract ("the caller reports false when it serves exact"). Unreachable today (no fleet storage is `AuthRecipeSQLStorage` without also being `ActivityLogStorage`), but the flag is inverted for the one branch that exists to serve exact.

---
Non-blocking awareness (not a bug — documented as intended): `deleteUserHelper` no longer scrubs `user_last_active` on delete, so an active-then-deleted user keeps counting until the row ages out of the window.
